File: dailies/16_aug_18/DFS.js
const GraphNode = require('../../helpers/GraphNode');

const depthFirstSearch = (root, id, visited = new Set()) => {
	if (!root) return undefined;
	else if (root.id === id) return root;

	visited.add(root.id);

	let n = undefined;

	for (const child of root.children) {
		if (!visited.has(child.id)) {
			n = depthFirstSearch(child, id, visited);
		}

		if (n) break;
	}

	return n;
};

const a = new GraphNode('a');
const b = new GraphNode('b');
const c = new GraphNode('c');
const d = new GraphNode('d');
const e = new GraphNode('e');
const f = new GraphNode('f');
const g = new GraphNode('g');
const h = new GraphNode('h');

a.addChild(d);
d.addChild(a);
a.addChild(c);
c.addChild(b);
b.addChild(f);
c.addChild(f);
f.addChild(h);
h.addChild(g);

console.log(depthFirstSearch(a, 'g'));

File: python/dailies/4_nov_22/dfs.py
from typing import Optional, Set
from helpers.graph_classes.graph_node import GraphNode
from helpers.testing.graph_search_tester import GraphSearchTester


def depth_first_search(
        root: Optional[GraphNode],
        identifier: int,
        visited: Optional[Set[GraphNode]] = None,
) -> Optional[GraphNode]:
    if root is None or root.get_id() == identifier:
        return root

    if visited is None:
        visited = set()

    visited.add(root)

    n = None

    for child in root.get_children():
        if child not in visited:
            n = depth_first_search(
                child, identifier, visited
            )

        if n is not None:
            break

    return n


GraphSearchTester(depth_first_search)
